rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,473 @@
+
+**Negative**
+
+- Nodes may have to go through millions of blocks requiring extra memory and
+computation over the transition phase
+- External clients would most likely need to support both block versions over
+the transition period
+- Bootstrapping nodes and light clients require an extra step for verification.
+- We've cornered off some untouchable fields in the block structure. If we
+really wanted to change these we would need a hard upgrade.
+- There is greater risk to correctness in the migration logic.
+
+## Status
+
+Proposed
+
+## References
+
+- [Upgrade tooling tracking issue](https://github.com/tendermint/tendermint/issues/5595)
+- [Protocol versions](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-016-protocol-versions.md)
+- [Go Modules: v2 and Beyond](https://blog.golang.org/v2-go-modules)
+
+## Appendix A: Upgrade terminology
+
+It's important to define upgrades and classify the distinctions so as to ensure
+that everyone has the same understanding of what it means and that we can use
+the same terminology to speak about the different concepts. This appendix aims
+at achieving this.
+
+An upgrade, in the loosest sense, is a change to the Tendermint codebase. For
+simplicity to users, we group changes together and form releases. Every release
+corresponds to a unique software version. Tendermint follows SemVer which
+has strict rules about versioning and makes a distinction between different
+types of releases.
+
+From the node operators perspective, upgrades just mean, stopping the node,
+changing the binary and starting the node again.
+
+A patch release (the last number in x.x.x), is a backwards compatible bug-fix.
+Nodes should be able to perform this upgrade at any time without any affect to
+the network.
+
+A minor release (the middle number in x.x.x), is also backwards compatible but
+it indicates changes in existing features or new features. One can think of
+performance optimizations as a good example. Nodes should also be able to
+perform this upgrade at any time and nodes with different minor versions should
+have no problem interacting with one another.
+
+We call the above two releases as constituting minor upgrades.
+
+A major release (the first number in x.x.x), is a set of incompatible changes.
+This means that the public api has changed and/or the messages that nodes send
+to one another and persist to disk have changed in a way that is incompatible
+with prior releases. A block protocol change is always a major release.
+
+When this happens nodes can't simply stop and restart when they want but must
+coordinate a restart together. There can't be nodes with different versions on
+the same network. Not only this, but upgraded nodes can't read nor verify data
+structures from a prior major version. Thus, an upgrade for a major release
+has so far meant having to create a new chain.
+
+Soft and hard upgrades both refer to major releases but have very different
```

Comment:
But I think all soft upgrades are major releases because external clients need to change their code in order to verify the new block format.

I guess my question here is whether an incompatible change constitutes only readability or also verifiability. If it's just about reading the blocks then maybe some soft upgrades could be released in a minor version